src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -2703,55 +2703,19 @@ protected override IPointerTypeSymbol CommonCreatePointerTypeSymbol(ITypeSymbol
             return CreatePointerTypeSymbol(elementType.EnsureCSharpSymbolOrNull<ITypeSymbol, TypeSymbol>("elementType"));
         }
 
-        protected override INamedTypeSymbol CommonCreateTupleTypeSymbol(ImmutableArray<ITypeSymbol> elementTypes, ImmutableArray<string> elementNames)
+        protected override INamedTypeSymbol CommonCreateTupleTypeSymbol(
+            ImmutableArray<ITypeSymbol> elementTypes,
+            ImmutableArray<string> elementNames,
+            ImmutableArray<Location> elementLocations)
         {
-            if (elementTypes.IsDefault)
-            {
-                throw new ArgumentNullException(nameof(elementTypes));
-            }
-
-            if (elementTypes.Length <= 1)
-            {
-                throw new ArgumentException(CodeAnalysisResources.TuplesNeedAtLeastTwoElements, nameof(elementNames));
-            }
-
-            elementNames = CheckTupleElementNames(elementTypes.Length, elementNames);
-
             var typesBuilder = ArrayBuilder<TypeSymbol>.GetInstance(elementTypes.Length);
             for (int i = 0; i < elementTypes.Length; i++)
             {
-                if (elementTypes[i] == null)
-                {
-                    throw new ArgumentNullException($"{nameof(elementTypes)}[{i}]");
-                }
-
                 typesBuilder.Add(elementTypes[i].EnsureCSharpSymbolOrNull<ITypeSymbol, TypeSymbol>($"{nameof(elementTypes)}[{i}]"));
             }
 
             return TupleTypeSymbol.Create(null, // no location for the type declaration
-                                          typesBuilder.ToImmutableAndFree(), default(ImmutableArray<Location>), elementNames, this);
-        }
-
-        /// <summary>
-        /// Check that if any names are provided, and their number matches the expected cardinality.
-        /// Returns a normalized version of the element names (empty array if all the names are null).
-        /// </summary>
-        private static ImmutableArray<string> CheckTupleElementNames(int cardinality, ImmutableArray<string> elementNames)
-        {
-            if (!elementNames.IsDefault)
-            {
-                if (elementNames.Length != cardinality)
-                {
-                    throw new ArgumentException(CodeAnalysisResources.TupleElementNameCountMismatch, nameof(elementNames));
-                }
-
-                if (elementNames.All(n => n == null))
-                {
-                    return default(ImmutableArray<string>);
-                }
-            }
-
-            return elementNames;
+                typesBuilder.ToImmutableAndFree(), elementLocations, elementNames, this);
```

Comment:
`typesBuilder` can be replaced with `var types = elementTypes.SelectAsArray((type, i, arg) => type.EnsureCSharpSymbolOrNull<ITypeSymbol, TypeSymbol>($"{nameof(elementTypes)}[i]"), (object)null);`